### ISSUE TITLE
Question: More details on the KV cache (HyperKV) variant from Appendix D?

### ISSUE THREAD

**Issue by qw3rtman** [SakanaAI/doc-to-lora#5]
Great work on D2L! I'm particularly interested in the KV cache generation variant described in Appendix D, where the hyper-network outputs prefix tokens instead of LoRA matrices.

A few questions:

1. How is the output head modified to produce K/V tensors instead of LoRA A/B matrices? Is it just swapping the projection dimensions, or were there other changes needed?

2. The appendix mentions using the same dataset as Section 4 (NIAH). Was the training objective also cross-entropy on ground-truth tokens, or did you use the forward KL loss from the main experiments?

3. Are there plans to evaluate HyperKV on the QA benchmarks (SQuAD, DROP, LongBench etc.)? Curious how it compares to the LoRA variant and to Cartridges on those tasks. I also wonder if you have any intuition on if this prefix-tuning meta-learning is easier or would maintain more performance over Cartridges, compared to D2L vs LoRA+forward KL.

4. Are the HyperKV checkpoint weights or training code available (or planned to be released)? I only see the LoRA variant on HuggingFace.

Thanks so much for the code + models release so far!

**Comment by 51616** [comment]
@qw3rtman Thank you for your interest in our project. Sorry for the late reply!

1. For the most part yes. We additionally remove the "decoder" part from the Perceiver architecture. Decoder is the part where the perceiver does cross-attention between latents.
2. We use CE with the GT tokens
3. We have no plan to run that experiment as we're actively working on the next phase of the project at the moment. Base on our internal experiments and Cartridges paper, I'd say that prefix training is better than LoRA for information retention purposes.
4. Yes you can download the checkpoints with
```bash
uv run huggingface-cli login
uv run huggingface-cli download SakanaAI/doc-to-lora --local-dir trained_d2l --include "*/"
```
